pkg/controller/controller_instance.go

```diff
@@ -1227,6 +1236,77 @@ func (c *controller) pollServiceInstance(serviceClass *v1alpha1.ServiceClass, se
 	return nil
 }
 
+// resolveReferences checks to see if ServiceClassRef and/or ServicePlanRef are
+// nil and if so, will resolve the references and update the instance.
+// If either can not be resolved, returns an error and sets the InstanceCondition
+// with the appropriate error message.
+func (c *controller) resolveReferences(instance *v1alpha1.ServiceInstance) (*v1alpha1.ServiceInstance, error) {
+	if instance.Spec.ServiceClassRef != nil && instance.Spec.ServicePlanRef != nil {
+		return instance, nil
+	}
+
+	if instance.Spec.ServiceClassRef == nil {
+		glog.V(4).Infof("looking up a ServiceClass from externalName: %q", instance.Spec.ExternalServiceClassName)
+		listOpts := apimachineryv1.ListOptions{FieldSelector: "spec.externalName==" + instance.Spec.ExternalServiceClassName}
+		serviceClasses, err := c.serviceCatalogClient.ServiceClasses().List(listOpts)
+		if err == nil && len(serviceClasses.Items) == 1 {
+			sc := &serviceClasses.Items[0]
+			instance.Spec.ServiceClassRef = &apiv1.ObjectReference{
+				Kind:            sc.Kind,
+				Namespace:       sc.Namespace,
+				Name:            sc.Name,
+				UID:             sc.UID,
+				APIVersion:      sc.APIVersion,
+				ResourceVersion: sc.ResourceVersion,
+			}
+			glog.V(4).Infof("Resolve ServiceClass %q to %q", instance.Spec.ExternalServiceClassName, instance.Spec.ServiceClassRef)
+		} else {
+			s := fmt.Sprintf("ServiceInstance \"%s/%s\" references a non-existent ServiceClass %q", instance.Namespace, instance.Name, instance.Spec.ExternalServiceClassName)
+			glog.Warning(s)
+			c.updateServiceInstanceCondition(
+				instance,
+				v1alpha1.ServiceInstanceConditionReady,
+				v1alpha1.ConditionFalse,
+				errorNonexistentServiceClassReason,
+				"The instance references a ServiceClass that does not exist. "+s,
+			)
+			c.recorder.Event(instance, api.EventTypeWarning, errorNonexistentServiceClassReason, s)
+			return nil, fmt.Errorf(s)
+		}
+	}
+
+	if instance.Spec.ServicePlanRef == nil {
```

Comment:
Actually, I think that we do plan adding support for switching plans in the future? In that case `ServicePlanRef` might not be null when the `ExternalServicePlanName` is updated. We will also need to check the `plan_updateable` flag for that case.
Currently it doesn't matter since we ignore any `Spec` updates, but that's a temporary hack. At least leaving a TODO comment there would be useful.

Reply:
Yes we do. I would assume that those kinds of validations might happen in validation class and TODO here might not be appropriate. If you feel strongly however, I can add it here.